src/Components/PowerBIWidget/Internal/PowerBIWidgetBuilder/PowerBIWidgetBuilder.tsx

```diff
@@ -0,0 +1,288 @@
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+import produce from 'immer';
+import {
+    IPowerBIWidgetBuilderProps,
+    IPowerBIWidgetBuilderStyleProps,
+    IPowerBIWidgetBuilderStyles
+} from './PowerBIWidgetBuilder.types';
+import { getStyles } from './PowerBIWidgetBuilder.styles';
+import {
+    classNamesFunction,
+    useTheme,
+    styled,
+    Stack,
+    TextField,
+    ComboBox
+} from '@fluentui/react';
+import { useTranslation } from 'react-i18next';
+import { getWidgetFormStyles } from '../../../ADT3DSceneBuilder/Internal/Behaviors/Widgets/WidgetForm/WidgetForm.styles';
+
+const getClassNames = classNamesFunction<
+    IPowerBIWidgetBuilderStyleProps,
+    IPowerBIWidgetBuilderStyles
+>();
+
+const PowerBIWidgetBuilder: React.FC<IPowerBIWidgetBuilderProps> = ({
+    styles,
+    formData,
+    adapter,
+    updateWidgetData,
+    setIsWidgetConfigValid
+}) => {
+    // contexts
+
+    // state
+    const [selectedType, setSelectedType] = useState(
+        formData?.widgetConfiguration?.type || 'visual'
+    );
+    const [reportUrl, setReportUrl] = useState(
+        formData?.widgetConfiguration?.embedUrl || ''
+    );
+    const [pageName, setPageName] = useState(
+        formData?.widgetConfiguration?.pageName || ''
+    );
+    const [, setVisualName] = useState(
+        formData?.widgetConfiguration?.visualName || ''
+    );
+
+    const [reportPages, setReportPages] = useState([]);
+    const disableReportPages = useMemo(() => {
+        return reportPages?.length < 1;
+    }, [reportPages]);
+    useEffect(() => {
+        if (!reportUrl) {
+            setReportPages([]);
+        }
+        adapter?.getPagesInReport(reportUrl).then((pages) => {
+            setReportPages(
+                pages?.map((page) => {
+                    return {
+                        key: page.name,
+                        text: page.displayName || page.name
+                    };
+                }) || []
+            );
+        });
+    }, [reportUrl, selectedType]);
+
+    const [reportVisuals, setVisuals] = useState([]);
+    const disableReportVisuals = useMemo(() => {
+        return reportVisuals?.length < 1;
+    }, [reportVisuals]);
+    useEffect(() => {
+        if (!reportUrl || !pageName) {
+            setVisuals([]);
+        }
+        adapter?.getVisualsOnPage(reportUrl, pageName).then((visuals) => {
+            setVisuals(
+                visuals?.map((visual) => {
```

Comment:
how sanitized is the data coming back from the adapter? Do we need any null or empty data checks or hardening in this logic?